### PR TITLE
TDT confidence fix

### PR DESCRIPTION
**Pull request by GNroy**
# What does this PR do ?

This PR enables confidence scores computing for TDT models.

It introduces a new `ConfidenceConfig` parameter named `tdt_include_duration` which allows caclulating frame-level confidence for the token duration prediction.

**Collection**: [ASR]

# Changelog 
- Add specific line by line info of high level changes in this PR.

# Usage
* You can potentially add a usage example below

```python
# Add a code snippet demonstrating how to use this 
```

# Jenkins CI

The Jenkins CI system has been replaced by GitHub Actions self-hosted runners.

There's no need to comment `jenkins` on the PR to trigger Jenkins CI.
The GitHub Actions CI will run automatically when the PR is opened.
To run CI on an untrusted fork, a NeMo user with write access must click "Approve and run".

# Before your PR is "Ready for review"
**Pre checks**:
- [ ] Make sure you read and followed [Contributor guidelines](https://github.com/NVIDIA/NeMo/blob/main/CONTRIBUTING.md)
- [ ] Did you write any new necessary tests?
- [ ] Did you add or update any necessary documentation?
- [ ] Does the PR affect components that are optional to install? (Ex: Numba, Pynini, Apex etc)
  - [ ] Reviewer: Does the PR have correct import guards for all optional libraries?
  
**PR Type**:
- [x] New Feature
- [x] Bugfix
- [ ] Documentation

If you haven't finished some of the above items you can still open "Draft" PR.

## Who can review?

Anyone in the community is free to review the PR once the checks have passed. 
[Contributor guidelines](https://github.com/NVIDIA/NeMo/blob/main/CONTRIBUTING.md) contains specific people who can review PRs to various areas.

# Additional Information
* Related to  https://github.com/NVIDIA/NeMo/issues/8737

### CODE REVIEW

File: nemo/collections/asr/parts/submodules/rnnt_decoding.py [hunk 1]
```diff
@@ -319,7 +330,7 @@ def __init__(self, decoding_cfg, decoder, joint, blank_id: int):
                         preserve_alignments=self.preserve_alignments,
                         preserve_frame_confidence=self.preserve_frame_confidence,
                         confidence_method_cfg=self.confidence_method_cfg,
-                        loop_labels=self.cfg.greedy.get('loop_labels', True),
+                        loop_labels=self.cfg.greedy.get('loop_labels', False),
```
[the remark this answers is not in the document]
Comment (GNroy) [reply nemo/collections/asr/parts/submodules/rnnt_decoding.py:333]: Rebase bug, will fix right away.

File: nemo/collections/asr/parts/submodules/rnnt_decoding.py [hunk 2]
```diff
@@ -334,8 +345,8 @@ def __init__(self, decoding_cfg, decoder, joint, blank_id: int):
                         ),
                         preserve_alignments=self.preserve_alignments,
                         preserve_frame_confidence=self.preserve_frame_confidence,
+                        include_duration_confidence=self.tdt_include_duration_confidence,
                         confidence_method_cfg=self.confidence_method_cfg,
-                        use_cuda_graph_decoder=self.cfg.greedy.get('use_cuda_graph_decoder', False),
```
[the remark this answers is not in the document]
Comment (GNroy) [reply nemo/collections/asr/parts/submodules/rnnt_decoding.py:338]: Same

File: nemo/collections/asr/parts/submodules/ctc_decoding.py [hunk 3]
```diff
@@ -98,6 +98,10 @@ class AbstractCTCDecoding(ConfidenceMixin):
                     Which aggregation type to use for collapsing per-token confidence into per-word confidence.
                     Valid options are `mean`, `min`, `max`, `prod`.
 
+                tdt_include_duration: Bool flag indicating that the duration confidence scores are to be calculated and
```
Comment (artbataev) [inline nemo/collections/asr/parts/submodules/ctc_decoding.py:101]: Why should it be in the CTC config (here and below)? I think in CTC models we do not need TDT confidence params

File: nemo/collections/asr/parts/submodules/rnnt_decoding.py [hunk 4]
```diff
@@ -569,28 +581,69 @@ def compute_confidence(self, hypotheses_list: List[Hypothesis]) -> List[Hypothes
...
+                                token_confidence.append(
+                                    self._aggregate_confidence(
+                                        [maybe_pre_aggregate(fc[i])] + subsequent_blank_confidence
+                                    )
                                 )
-                            )
-                            offset = 0
+                                subsequent_blank_confidence = []
+                    else:
+                        i, a = 0, fa[0]
```
Comment (artbataev) [inline nemo/collections/asr/parts/submodules/rnnt_decoding.py:611]: Please, add comments what's going on here and why

**Review by artbataev — CHANGES_REQUESTED** [review]
I found at least 2 problems with the current implementation:
1) when decoding with duration confidence, inside `GreedyBatchedTDTLoopLabelsComputer`  `include_duration_confidence` parameter is False (you can print it)
```shell
python examples/asr/speech_to_text_eval.py \
 pretrained_name=nvidia/parakeet-tdt-1.1b \
 dataset_manifest=~/code/asr_particles/manifests/librispeech/test_other.json \
 batch_size=16 \
 output_filename=test_other_decoded.jsonl \
 amp=false \
 amp_dtype=bfloat16 \
 rnnt_decoding.greedy.tdt_include_duration_confidence=true \
 rnnt_decoding.greedy.preserve_frame_confidence=true \
 preserve_alignment=true
```
- I think that also you need to adjust `BatchedAlignments._allocate_more` method to correctly reflect the changed shape `self.frame_confidence = torch.cat((self.frame_confidence, torch.zeros_like(self.frame_confidence)), dim=1)` (should be `dim=1`)

2) the model fails with bf16 (`amp=true`)
```shell
python examples/asr/speech_to_text_eval.py \
 pretrained_name=nvidia/parakeet-tdt-1.1b \
 dataset_manifest=~/code/asr_particles/manifests/librispeech/test_other.json \
 batch_size=16 \
 output_filename=test_other_decoded.jsonl \
 amp=true \
 amp_dtype=bfloat16 \
 rnnt_decoding.greedy.tdt_include_duration_confidence=true \
 rnnt_decoding.greedy.preserve_frame_confidence=true \
 preserve_alignment=true
```

File: nemo/collections/asr/parts/submodules/tdt_loop_labels_computer.py [hunk 5]
```diff
@@ -327,7 +334,19 @@ def loop_labels_torch(
                     time_indices=time_indices_current_labels,
                     logits=logits if self.preserve_alignments else None,
                     labels=labels if self.preserve_alignments else None,
-                    confidence=self._get_confidence_tensor(F.log_softmax(logits[:, :-num_durations], dim=-1))
+                    confidence=torch.cat(
```
Comment (artbataev) [inline nemo/collections/asr/parts/submodules/tdt_loop_labels_computer.py:337]: Use `torch.stack` to concat along new dimension (instead of multiple unsqueeze)

File: nemo/collections/asr/parts/submodules/ctc_decoding.py (hunk 3 above)
Comment (GNroy) [reply nemo/collections/asr/parts/submodules/ctc_decoding.py:101]: To my understanding, comments in ctc_decoding and rnnt_decoding just display the content of the confidence config.

File: nemo/collections/asr/parts/submodules/rnnt_decoding.py (hunk 4 above)
Comment (GNroy) [reply nemo/collections/asr/parts/submodules/rnnt_decoding.py:611]: There is a comment a few lines above `# there is only one score per frame most of the time`.
Apparently, it's not enough. The idea is to skip running `for` loop for frames with a single prediction.

File: nemo/collections/asr/parts/submodules/tdt_loop_labels_computer.py (hunk 5 above)
Comment (GNroy) [reply nemo/collections/asr/parts/submodules/tdt_loop_labels_computer.py]: Thanks, done.

**Comment by GNroy** [comment]
> I found 2 problems with the current implementation:

I think I fixed those problems. Could you please check again?

**Review by artbataev — DISMISSED** [review]
LGTM, thanks!

**Review by artbataev — APPROVED** [review]
